src/transactions/BumpExpirationOpFrame.cpp

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+#include "transactions/BumpExpirationOpFrame.h"
+
+namespace stellar
+{
+
+struct BumpExpirationMetrics
+{
+    medida::MetricsRegistry& mMetrics;
+
+    size_t mLedgerWriteByte{0};
+
+    BumpExpirationMetrics(medida::MetricsRegistry& metrics) : mMetrics(metrics)
+    {
+    }
+
+    ~BumpExpirationMetrics()
+    {
+        mMetrics
+            .NewMeter({"soroban", "bump-exp-op", "write-ledger-byte"}, "byte")
+            .Mark(mLedgerWriteByte);
+    }
+};
+
+BumpExpirationOpFrame::BumpExpirationOpFrame(Operation const& op,
+                                             OperationResult& res,
+                                             TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mBumpExpirationOp(mOperation.body.bumpExpirationOp())
+{
+}
+
+ThresholdLevel
+BumpExpirationOpFrame::getThresholdLevel() const
+{
+    return ThresholdLevel::LOW;
+}
+
+bool
+BumpExpirationOpFrame::isOpSupported(LedgerHeader const& header) const
+{
+    return header.ledgerVersion >= 20;
+}
+
+bool
+BumpExpirationOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    throw std::runtime_error("BumpExpirationOpFrame::doApply needs Config");
+}
+
+bool
+BumpExpirationOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
+                               Hash const& sorobanBasePrngSeed)
+{
+    BumpExpirationMetrics metrics(app.getMetrics());
+
+    auto const& resources = mParentTx.sorobanResources();
+    auto const& footprint = resources.footprint;
+
+    UnorderedMap<LedgerKey, uint32_t> originalExpirations;
+
+    for (auto const& lk : footprint.readOnly)
+    {
+        // When we move to use EXPIRATION_EXTENSIONS, this should become a
+        // loadWithoutRecord, and the metrics should be updated.
+        auto ltxe = ltx.load(lk);
+        if (ltxe)
+        {
+            auto keySize = xdr::xdr_size(lk);
+            auto entrySize = xdr::xdr_size(ltxe.current());
+            metrics.mLedgerWriteByte += keySize;
+            metrics.mLedgerWriteByte += entrySize;
+
+            // Multiply by 2 for the metadata check since the previous state is
+            // also included in the meta.
+            if (resources.extendedMetaDataSizeBytes <
+                    metrics.mLedgerWriteByte * 2 ||
```

Comment:
To avoid integer overflow with multiplication of 2, should we use division instead?

Reply:
Done. This should be changing anyways when we start using EXPIRATION_EXTENSIONS.